File: frontend/ai-client/src/core/network/api_client.dart
/// Authenticated API Client for Flutter
///
/// Features:
/// - Automatic token injection
/// - Token refresh on 401
/// - Tenant header injection

import 'package:dio/dio.dart';
import '../auth/auth_storage.dart';
import '../auth/jwt_utils.dart';

class AuthenticatedApiClient {
  final Dio _dio;
  final AuthStorage _storage;
  final String _authBaseUrl;

  bool _isRefreshing = false;
  final List<Function(String)> _pendingRequests = [];

  AuthenticatedApiClient({
    required String baseUrl,
    required AuthStorage storage,
    String? authBaseUrl,
  })  : _storage = storage,
        _authBaseUrl = authBaseUrl ?? 'http://localhost:8001',
        _dio = Dio(BaseOptions(
          baseUrl: baseUrl,
          connectTimeout: const Duration(seconds: 30),
          receiveTimeout: const Duration(seconds: 120),
          headers: {
            'Content-Type': 'application/json',
          },
        )) {
    _setupInterceptors();
  }

  Dio get client => _dio;

  void _setupInterceptors() {
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: _onRequest,
      onError: _onError,
    ));
  }

  Future<void> _onRequest(
    RequestOptions options,
    RequestInterceptorHandler handler,
  ) async {
    // Get and validate token
    String? token = await _storage.getAccessToken();

    if (token != null) {
      // Check if token needs refresh
      if (JwtUtils.isExpired(token, bufferSeconds: 60)) {
        token = await _refreshToken();
      }

      if (token != null) {
        options.headers['Authorization'] = 'Bearer $token';
      }
    }

    // Add tenant header
    final tenantId = await _storage.getTenantId();
    if (tenantId != null) {
      options.headers['X-Tenant-ID'] = tenantId;
    }

    handler.next(options);
  }

  Future<void> _onError(
    DioException error,
    ErrorInterceptorHandler handler,
  ) async {
    if (error.response?.statusCode == 401) {
      // Token expired - try refresh
      final newToken = await _refreshToken();

      if (newToken != null) {
        // Retry original request
        final opts = error.requestOptions;
        opts.headers['Authorization'] = 'Bearer $newToken';

        try {
          final response = await _dio.fetch(opts);
          return handler.resolve(response);
        } catch (e) {
          return handler.reject(error);
        }
      }
    }

    handler.next(error);
  }

  Future<String?> _refreshToken() async {
    if (_isRefreshing) {
      // Wait for ongoing refresh
      return Future<String?>((resolve) {
        _pendingRequests.add(resolve as Function(String));
      });
    }

    _isRefreshing = true;

    try {
      final response = await Dio().post(
        '$_authBaseUrl/auth/refresh',
        options: Options(
          headers: {'Content-Type': 'application/json'},
        ),
      );

      final newToken = response.data['access_token'] as String?;

      if (newToken != null) {
        await _storage.saveAccessToken(newToken);

        // Resolve pending requests
        for (final callback in _pendingRequests) {
          callback(newToken);
        }
        _pendingRequests.clear();

        return newToken;
      }
    } catch (e) {
      // Refresh failed - clear tokens
      await _storage.clearAll();
    } finally {
      _isRefreshing = false;
    }

    return null;
  }

  // Convenience methods

  Future<Response<T>> get<T>(
    String path, {
    Map<String, dynamic>? queryParameters,
  }) {
    return _dio.get<T>(path, queryParameters: queryParameters);
  }

  Future<Response<T>> post<T>(
    String path, {
    dynamic data,
  }) {
    return _dio.post<T>(path, data: data);
  }

  Future<Response<T>> put<T>(
    String path, {
    dynamic data,
  }) {
    return _dio.put<T>(path, data: data);
  }

  Future<Response<T>> delete<T>(String path) {
    return _dio.delete<T>(path);
  }
}

/// RAG API Client
class RagApiClient extends AuthenticatedApiClient {
  RagApiClient({
    required AuthStorage storage,
  }) : super(
          baseUrl: const String.fromEnvironment(
            'RAG_API_URL',
            defaultValue: 'http://localhost:8009',
          ),
          storage: storage,
        );
}

File: frontend/web-console/core/api-client.ts
/**
 * Secure API Client with Token Management
 * 
 * Features:
 * - Access token in memory only
 * - Automatic refresh before expiry
 * - Retry failed requests after refresh
 * - Tenant header injection
 */

import axios, {
    AxiosInstance,
    AxiosError,
    InternalAxiosRequestConfig,
    AxiosResponse,
} from 'axios';
import { isTokenExpired, getUserFromToken, getTokenTimeRemaining } from './jwt';

const AUTH_API = process.env.NEXT_PUBLIC_AUTH_API || 'http://localhost:8001';
const RAG_API = process.env.NEXT_PUBLIC_RAG_API || 'http://localhost:8009';
const PIPELINE_API = process.env.NEXT_PUBLIC_PIPELINE_API || 'http://localhost:8008';
const INGESTION_API = process.env.NEXT_PUBLIC_INGESTION_API || 'http://localhost:8002';

// Export API endpoints for use in services
export const API_ENDPOINTS = {
    auth: AUTH_API,
    rag: RAG_API,
    pipeline: PIPELINE_API,
    ingestion: INGESTION_API,
};

/**
 * Token storage (memory only - no localStorage for security)
 */
class TokenManager {
    private accessToken: string | null = null;
    private refreshPromise: Promise<string | null> | null = null;
    private refreshTimer: NodeJS.Timeout | null = null;

    setAccessToken(token: string): void {
        this.accessToken = token;
        this.scheduleRefresh(token);
    }

    getAccessToken(): string | null {
        return this.accessToken;
    }

    clearTokens(): void {
        this.accessToken = null;
        if (this.refreshTimer) {
            clearTimeout(this.refreshTimer);
            this.refreshTimer = null;
        }
    }

    getTenantId(): string | null {
        if (!this.accessToken) return null;
        const user = getUserFromToken(this.accessToken);
        return user?.tenantId || null;
    }

    isAuthenticated(): boolean {
        return this.accessToken !== null && !isTokenExpired(this.accessToken);
    }

    private scheduleRefresh(token: string): void {
        if (this.refreshTimer) {
            clearTimeout(this.refreshTimer);
        }

        const remaining = getTokenTimeRemaining(token);
        // Refresh 60 seconds before expiry
        const refreshIn = Math.max(0, (remaining - 60) * 1000);

        this.refreshTimer = setTimeout(() => {
            this.performRefresh();
        }, refreshIn);
    }

    async performRefresh(): Promise<string | null> {
        // Prevent concurrent refresh attempts
        if (this.refreshPromise) {
            return this.refreshPromise;
        }

        this.refreshPromise = this.doRefresh();
        const result = await this.refreshPromise;
        this.refreshPromise = null;
        return result;
    }

    private async doRefresh(): Promise<string | null> {
        try {
            const response = await axios.post(
                `${AUTH_API}/auth/refresh`,
                {},
                { withCredentials: true }
            );

            const newToken = response.data.access_token;
            if (newToken) {
                this.setAccessToken(newToken);
                return newToken;
            }
        } catch (error) {
            this.clearTokens();
        }
        return null;
    }
}

export const tokenManager = new TokenManager();

/**
 * Create authenticated API client
 */
function createAuthenticatedClient(baseURL: string): AxiosInstance {
    const client = axios.create({
        baseURL,
        timeout: 30000,
        withCredentials: true,
        headers: {
            'Content-Type': 'application/json',
        },
    });

    // Request interceptor - inject tokens
    client.interceptors.request.use(
        async (config: InternalAxiosRequestConfig) => {
            let token = tokenManager.getAccessToken();

            // Check if token needs refresh
            if (token && isTokenExpired(token, 60)) {
                token = await tokenManager.performRefresh();
            }

            if (token) {
                config.headers.Authorization = `Bearer ${token}`;
            }

            const tenantId = tokenManager.getTenantId();
            if (tenantId) {
                config.headers['X-Tenant-ID'] = tenantId;
            }

            return config;
        },
        (error) => Promise.reject(error)
    );

    // Response interceptor - handle 401 and retry
    client.interceptors.response.use(
        (response: AxiosResponse) => response,
        async (error: AxiosError) => {
            const originalRequest = error.config as InternalAxiosRequestConfig & {
                _retry?: boolean;
            };

            if (error.response?.status === 401 && !originalRequest._retry) {
                originalRequest._retry = true;

                const newToken = await tokenManager.performRefresh();
                if (newToken) {
                    originalRequest.headers.Authorization = `Bearer ${newToken}`;
                    return client(originalRequest);
                }

                // Refresh failed - redirect to login
                if (typeof window !== 'undefined') {
                    window.location.href = '/login?session_expired=true';
                }
            }

            return Promise.reject(error);
        }
    );

    return client;
}

// Export configured clients
export const authApi = createAuthenticatedClient(AUTH_API);
export const ragApi = createAuthenticatedClient(RAG_API);
export const pipelineApi = createAuthenticatedClient(PIPELINE_API);

// Public client for login (no auth required)
export const publicAuthApi = axios.create({
    baseURL: AUTH_API,
    timeout: 30000,
    withCredentials: true,
    headers: {
        'Content-Type': 'application/json',
    },
});
